src/python/pants/backend/java/compile/javac.py

```diff
@@ -60,26 +44,53 @@ async def compile_java_source(
     javac_binary: JavacBinary,
     request: CompileJavaSourceRequest,
 ) -> CompiledClassfiles:
-    sources = await Get(
-        SourceFiles,
-        SourceFilesRequest(
-            (request.target.get(Sources),),
-            for_sources_types=(JavaSources,),
-            enable_codegen=True,
+    component_members_with_sources = tuple(
+        t for t in request.component.members if t.has_field(Sources)
+    )
+    component_members_and_source_files = zip(
+        component_members_with_sources,
+        await MultiGet(
+            Get(
+                SourceFiles,
+                SourceFilesRequest(
+                    (t.get(Sources),),
+                    for_sources_types=(JavaSources,),
+                    enable_codegen=True,
+                ),
+            )
+            for t in component_members_with_sources
         ),
     )
-    if sources.snapshot.digest == EMPTY_DIGEST:
+
+    component_members_and_java_source_files = [
+        (target, sources)
+        for target, sources in component_members_and_source_files
+        if sources.snapshot.digest != EMPTY_DIGEST
+    ]
+
+    if not component_members_and_java_source_files:
         return CompiledClassfiles(digest=EMPTY_DIGEST)
 
-    lockfile, direct_deps = await MultiGet(
-        Get(CoursierResolvedLockfile, CoursierLockfileForTargetRequest(Targets((request.target,)))),
-        Get(Targets, DependenciesRequest(request.target.get(Dependencies))),
+    # Target coarsening currently doesn't perform dep expansion, which matters for targets
+    # with multiple sources that expand to individual source subtargets.
+    # We expand the dependencies explicitly here before coarsening, but ideally this could
+    # be done somehow during coarsening.
+    # TODO: Should component dependencies be filtered out here if they were only brought in by component members which were
+    #   filtered out above (due to having no JavaSources to contribute)?  If so, that will likely required extending
+    #   the CoarsenedTargets API to include more complete dependency information, or to support such filtering directly.
+    expanded_direct_deps = await Get(Targets, Addresses(request.component.dependencies))
+    coarsened_direct_deps = await Get(
+        CoarsenedTargets, Addresses(t.address for t in expanded_direct_deps)
     )
```

Comment:
You can see how the graph ends up being complete [in the unit tests](https://github.com/pantsbuild/pants/blob/b4febf1c392aa05269af782401366e030e38a6e3/src/python/pants/engine/internals/graph_test.py#L338-L383): the base targets (dep, t1, and t2) depend on their own files.

Reply:
Yeah, I had a fair amount of trouble deciding on an overall mental model here.  Consuming a `CoarsenedTarget` in `CompileJavaSourceRequest` was where I landed.  I also waffled for quite a while on spliting this into two rules: one for dealing with expanding and coarsening targets, and one lower level rule for actually running the compilation against a set of `JavaSources` fields.

I'm open to changing this, but let me explain my thought process:

> if you're compiling recursively, some nodes in the recursion not having sources shouldn't short-circuit the recursion.

The whole crux of the problem with `javac` is limiting my dependency classpath to only the current compilation unit's **direct** dependencies.  We don't want the full transitive closure on the classpath of the current compilation unit, by design, even if we might indirectly, recursively build it in order to build our direct dependencies.  So we're really distinguishing here between our compile-time (direct dep) classpath, and a "runtime" classpath that is likely the closure's classpath.  This is why the recursion eagerly terminates when it concludes that the current component doesn't produce any classpath entries: we don't need to build it, which means we don't need its deps (only used to build it), which means we don't recursively compile its deps.

But by direct dependency, we really mean direct **source** dependency, or possibly some other kind of resource dependency that can somehow contribute to the classpath but isn't yet modeled.  This is where things get sticky, because target expansion sort of changes the definition of "direct" dependency.

> You can see how the graph ends up being complete in the unit tests: the CoarsenedTargets for the base targets (dep, t1, and t2) depend on their own files.

That test only uses explicit file dependencies; it doesn't cover the case I'm covering here where, e.g. `t1` depends on `dep` instead of `dep.txt:dep`.

An example:

`A.java` -> `B` -- (expands to) --> (`B.java` and `B2.java`).

If we use target expansion, we'll automatically get `A.java` -> `B.java`;  `A.java` -> `C.java`; and `B.java` <--> `C.java`.  So subtarget expansion ends up getting us our natural notion of direct dependencies.  Just coarsening `B` will not do this; we need to do both.  See `javac_test.py:test_compile_with_transitive_multiple_sources` in this PR for a concrete example of this.

Stepping way back: my mental model here was to operate over the nodes of an induced **coarsened, expanded** graph.  Then each rule invocation maps 1:1 with a coarsened, expanded component, and its only job is to make sure that its recursive calls are also over coarsened, expanded components.  Expanding gets us the right notion of direct dependencies in the presence of subtarget expansion, while coarsening gets us the right notion of a single compilation unit in the presence of source-target cycles.

I'm pretty confident this needs a more generic model, since we're going to need to eventually worry about other rules that can contribute to the compilation classpath, which means we're probably going to need to request a union product from our dependencies.  That's where I was imagining two layers of rules, where `compile_java_source` is lower level and not expected to think hard about how to expand its dependencies.  But without another example of a rule that produces files for the compilation classpath, I worry that attempting to model that is premature.  So I opted for the "direct" route here just to get `javac` working and exercising the existing coarsening/expansion machinery.

Again, I'm happy to change this if you have an idea of how to remodel it, but pay particular attention to the example in `javac_test.py:test_compile_with_transitive_multiple_sources` to make sure that that case is covered (or why that case is somehow not actually a valid graph to test against).